boa/src/object/gcobject.rs

```diff
@@ -199,48 +206,57 @@ impl GcObject {
     /// Panics if the object is currently mutably borrowed.
     // <https://tc39.es/ecma262/#sec-ecmascript-function-objects-construct-argumentslist-newtarget>
     #[track_caller]
-    pub fn construct(&self, args: &[Value], context: &mut Context) -> Result<Value> {
-        // If the prototype of the constructor is not an object, then use the default object
-        // prototype as prototype for the new object
-        // see <https://tc39.es/ecma262/#sec-ordinarycreatefromconstructor>
-        // see <https://tc39.es/ecma262/#sec-getprototypefromconstructor>
-        let proto = self.get(&PROTOTYPE.into(), context)?;
-        let proto = if proto.is_object() {
-            proto
-        } else {
-            context
-                .standard_objects()
-                .object_object()
-                .prototype()
-                .into()
-        };
-        let this: Value = Object::create(proto).into();
-
+    pub fn construct(
+        &self,
+        args: &[Value],
+        new_target: Value,
+        context: &mut Context,
+    ) -> Result<Value> {
         let this_function_object = self.clone();
         let body = if let Some(function) = self.borrow().as_function() {
             if function.is_constructable() {
                 match function {
                     Function::BuiltIn(BuiltInFunction(function), _) => {
-                        FunctionBody::BuiltIn(*function)
+                        FunctionBody::BuiltInConstructor(*function)
                     }
                     Function::Ordinary {
                         body,
                         params,
                         environment,
                         flags,
                     } => {
+                        // If the prototype of the constructor is not an object, then use the default object
+                        // prototype as prototype for the new object
+                        // see <https://tc39.es/ecma262/#sec-ordinarycreatefromconstructor>
+                        // see <https://tc39.es/ecma262/#sec-getprototypefromconstructor>
+                        let proto = new_target
+                            .as_object()
+                            .unwrap()
+                            .get(&PROTOTYPE.into(), context)?;
+                        let proto = if proto.is_object() {
+                            proto
+                        } else {
+                            context
+                                .standard_objects()
+                                .object_object()
+                                .prototype()
+                                .into()
+                        };
+                        let this: Value = Object::create(proto).into();
```

Comment:
I usually tend to use `From` when I need an explicit conversion to a given type. It seems more clear, and doesn't need type hints:
```suggestion
                        let this = Value::from(Object::create(proto));
```